### PR TITLE
Pe local schur reorg

### PR DESCRIPTION
**Pull request by Paulie14**
Implementation of 2. Schur complement assembly for Darcy LMH and Richards.

- not supporting BDDC in LMH
- not supporting Mortar methods in LMH

TESTS:
- 11_darcy_bc/19_seepage_stagnation_LMH test fails - different results, however it also ends with stagnation (which is actually tested). [UPDATED]
- 12 - mortar LMH tests should be skipped, [SKIPPED]

### CODE REVIEW

File: src/flow/assembly_lmh.hh [hunk 1]
```diff
@@ -53,46 +52,50 @@ public:
...
         // sp.diag(-nsides-1).ones();
         // sp.print();
         
-        sp.submat(0, nsides+1, nsides-1, size()-1).diag().ones();
-        sp.submat(nsides+1, 0, size()-1, nsides-1).diag().ones();
+        base_local_sp_.submat(0, nsides+1, nsides-1, size()-1).diag().ones();
+        base_local_sp_.submat(nsides+1, 0, size()-1, nsides-1).diag().ones();
         
-        loc_system_.set_sparsity(sp);
+        loc_system_.set_sparsity(base_local_sp_);
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/flow/assembly_lmh.hh]: Full sparsity pattern as default.

File: src/flow/assembly_lmh.hh [hunk 2]
```diff
@@ -174,37 +270,171 @@ public:
...
+                    dofs_schur[t] = cr_dofs[j];
+                    break;
+                }
+            }
+            i++;
+        }
+    }
     
-    void set_dofs_and_bc(LocalElementAccessorBase<3> ele_ac){
+    void set_dofs(LocalElementAccessorBase<3> ele_ac){
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/flow/assembly_lmh.hh]: Issue to use cell loc dof getter returning arma::uvec and implement it.

File: src/flow/assembly_lmh.hh [hunk 3]
```diff
@@ -435,29 +660,31 @@ protected:
...
-                time_term_rhs = time_term_diag * ad_->previous_solution[ele_ac.edge_local_row(i)];
+                time_term_rhs = time_term_diag * ad_->previous_time_schur_solution[loc_schur_.row_dofs[i]];
             }
 
             this->loc_system_.add_value(loc_edge_dofs[i], loc_edge_dofs[i],
                                         -time_term_diag,
                                         -source_term - time_term_rhs);
-                
+            
+            if( ! reconstruct)
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/flow/assembly_lmh.hh]: Possibly pass it as explictit parameter to assembly_bc.

File: src/coupling/balance.cc [hunk 4]
```diff
@@ -182,7 +183,19 @@ void Balance::allocate(unsigned int n_loc_dofs,
...
+
+void Balance::allocate(const std::shared_ptr<DOFHandlerMultiDim>& dh,
+		unsigned int max_dofs_per_boundary)
+{
+    ASSERT(! allocation_done_);
+	// for sequential matrices, we need to include ghost values
+    n_loc_dofs_seq_ = dh->get_local_to_global_map().size();
+	// for parallel matrices, we use the local size from dof distribution
+	n_loc_dofs_par_ = dh->distr()->lsize();
     max_dofs_per_boundary_ = max_dofs_per_boundary;
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/coupling/balance.cc]: Issue #1154

File: src/flow/assembly_lmh.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Issue #1155.

File: src/flow/assembly_lmh.hh [hunk 5]
```diff
@@ -104,57 +107,150 @@ public:
...
+    //         schur_solution(i) = ad_->schur_solution[dofs_schur[i]];
+    //     }
+        
+    //     // reconstruct the velocity and pressure
+    //     ls.reconstruct_solution_schur(schur_offset_, schur_solution, reconstructed_solution_);
+
+    //     postprocess_velocity(ele_ac.dh_cell(), reconstructed_solution_);
+    // }
+
+    void load_local_system(const DHCellAccessor& dh_cell)
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Documented c19a91ea0d70190c62fa0eba4be99d69f02cfd59.

File: src/flow/assembly_lmh.hh [hunk 6]
```diff
@@ -104,57 +107,150 @@ public:
...
+        // possibly find the corresponding local system
+        auto ls = ad_->seepage_bc_systems.find(dh_cell.elm_idx());
+        if (ls != ad_->seepage_bc_systems.end())
+        {
+            const unsigned int n = loc_schur_.row_dofs.n_rows;  // local Schur complement size
+            arma::vec schur_solution(n);
+            for(unsigned int i=0; i<n; i++)
+            {
+                // read the computed edge pressures
+                schur_solution(i) = ad_->schur_solution[loc_schur_.row_dofs[i]];
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: I suppose, the function ```get_subvec``` should only read the vector from row_dofs[0] to row_dofs.size()-1, considering continuous dof indices.
Or should it go generally through the given row_dofs and pick up the corresponding elements?

File: src/flow/assembly_lmh.hh [hunk 7]
```diff
@@ -174,37 +270,171 @@ public:
...
+
 protected:
-    
     static const unsigned int size()
     {
         // dofs: velocity, pressure, edge pressure
         return RefElement<dim>::n_sides + 1 + RefElement<dim>::n_sides;
     }
+
+    void set_loc_dofs_vec(LocalElementAccessorBase<3> ele_ac, arma::Col<LongIdx>& dofs, arma::Col<LongIdx>& dofs_schur){
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Issue #1156.

File: src/flow/assembly_lmh.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Issue #1156.

File: src/flow/assembly_lmh.hh [hunk 8]
```diff
@@ -517,10 +748,22 @@ protected:
...
+                              * ad_->cross_section.value(ele.centre(), ele)
+                              / ele->n_sides();
+        
+        double edge_source_term = edge_scale * ad_->water_source_density.value(ele.centre(), ele);
+      
+        postprocess_velocity_specific(dh_cell, solution, edge_scale, edge_source_term);
+    }
 
-    virtual void postprocess_velocity_specific(const DHCellAccessor& dh_cell, double edge_scale, double edge_source_term)// override
+    virtual void postprocess_velocity_specific(const DHCellAccessor& dh_cell, arma::vec& solution,
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Issue #1157.

File: src/flow/darcy_flow_lmh.cc [hunk 9]
```diff
@@ -181,8 +181,8 @@ DarcyLMH::DarcyLMH(Mesh &mesh_in, const Input::Record in_rec)
     output_object(nullptr),
     solution(nullptr),
     data_changed_(false),
-    schur0(nullptr),
-	par_to_all(nullptr)
+    schur_compl(nullptr),
+    par_to_all(nullptr)
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/darcy_flow_lmh.cc]: Done in PE_Richards_out

File: src/flow/darcy_flow_lmh.cc [hunk 10]
```diff
@@ -347,19 +348,55 @@ void DarcyLMH::initialize() {
...
     initialize_specific();
 }
 
 void DarcyLMH::initialize_specific()
-{}
+{
+    data_->multidim_assembler = AssemblyBase::create< AssemblyLMH >(data_);
+}
+
+// void DarcyLMH::read_initial_condition()
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/darcy_flow_lmh.cc]: Issue #1158.

File: src/flow/darcy_flow_lmh.hh [hunk 11]
```diff
@@ -153,14 +154,19 @@ public:
...
-        VectorMPI previous_solution;
-        
         // Propagate test for the time term to the assembly.
         // This flag is necessary for switching BC to avoid setting zero neumann on the whole boundary in the steady case.
         bool use_steady_assembly_;
         
         // for time term assembly
         double time_step_;
+        
+        LinSys *lin_sys_schur;
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/darcy_flow_lmh.hh]: Done in PE_Richards_out.

File: src/flow/darcy_flow_lmh.hh [hunk 12]
```diff
@@ -252,13 +258,15 @@ protected:
...
+    
+    void reconstruct_solution_from_schur(MultidimAssembly& assembler);
 
     /**
      * Assembly or update whole linear system.
      */
     virtual void assembly_linear_system();
 
-    void set_mesh_data_for_bddc(LinSys_BDDC * bddc_ls);
+//     void set_mesh_data_for_bddc(LinSys_BDDC * bddc_ls);
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/darcy_flow_lmh.hh]: Issue #1153.

File: tests/12_darcy_frac/config.yaml [hunk 13]
```diff
@@ -11,7 +11,7 @@ test_cases:
       - 06_cube.yaml
       - 52_vtk_LMH.yaml
       - 53_piezo_LMH.yaml
-      - 54_bddc_LMH.yaml
+      #- 54_bddc_LMH.yaml
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply tests/12_darcy_frac/config.yaml]: Issue #1153.

File: src/la/vector_mpi.hh [hunk 14]
```diff
@@ -104,8 +104,11 @@ public:
         chkerr(VecZeroEntries( data_petsc_ ));
     }
 
-    /// Return new vector with same parallel structure.
-    void duplicate(VectorMPI other) {
+    /// For the current vector, it creates the same parallel structure as the @p other vector has.
+    /**
+     * FIXME: it does not take care of ghost values, so it cannot be used in that situation.
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/la/vector_mpi.hh]: Issue for VectorMPI Improvement: #1092.

File: src/la/local_system.hh [hunk 15]
```diff
@@ -137,6 +133,24 @@ public:
      * Almost_zero values will be set in all entries: sp(i,j) != 0
      */
     void set_sparsity(const arma::umat & sp);
+    
+    /** @brief Computes Schur complement of the local system: S = C - B * invA * Bt
+     * Applicable for square matrices.
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/la/local_system.hh]: Done in 0aa1a027443a289b080a300f95ed648145a090a6.

File: src/flow/richards_lmh.cc [hunk 16]
```diff
@@ -133,28 +133,32 @@ void RichardsLMH::initialize_specific() {
...
+        data_->multidim_assembler[dh_cell.elm().dim()-1]->update_water_content(dh_cell);
+    }
 }
 
 
 void RichardsLMH::prepare_new_time_step()
 {
-    data_->previous_solution.copy(data_->data_vec_);
-    data_->water_content_previous_time.copy(data_->water_content_previous_it);
+    data_->previous_time_schur_solution.copy_from(data_->schur_solution);
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/richards_lmh.cc]: Done in e9277324400ed5cd26b6913d357ee1e354c5d12c.

File: src/flow/richards_lmh.cc [hunk 17]
```diff
@@ -133,28 +133,32 @@ void RichardsLMH::initialize_specific() {
...
+    // modify side fluxes in parallel
+    // for every local edge take time term on diagonal and add it to the corresponding flux
+    
+    for ( DHCellAccessor dh_cell : data_->dh_->own_range() ) {
+        data_->multidim_assembler[dh_cell.elm().dim()-1]->update_water_content(dh_cell);
+    }
 }
 
 
 void RichardsLMH::prepare_new_time_step()
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/richards_lmh.cc]: Done in ee4aa3dc5879925c8c7cddb9d800b17942353957.

File: src/flow/darcy_flow_lmh.cc [hunk 18]
```diff
@@ -411,18 +454,21 @@ void DarcyLMH::zero_time_step()
     // zero_time_term means steady case
     data_->use_steady_assembly_ = zero_time_term();
 
-    data_->data_vec_.zero_entries();
+
+    data_->full_solution.zero_entries();
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/darcy_flow_lmh.cc]: Removed in be09ef42a5bb38a88c8c6242234b60184b9dd477.

File: src/flow/assembly_lmh.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: generally, not considering continuous vector..

File: src/flow/assembly_lmh.hh [hunk 19]
```diff
@@ -174,37 +270,171 @@ public:
...
+        
+        // TODO: sparse for commmunication blocks?
+        schur_sp_.ones(loc_size_schur, loc_size_schur);
+        
+       
+        arma::Col<LongIdx> d = indices;
+        dofs_schur.head(indices.size()) = d;
+        
+        //D, E',E block: compatible connections: element-edge
+        Neighbour *ngh;
```
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Done by 1afeb8e4affb62783dba5eea6c73c82b02e47f76.

File: src/flow/assembly_lmh.hh (hunk 3 above)
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Done. And for balance created issue #1160.

File: src/flow/assembly_lmh.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Done as a part of issue #1092.

**Pull request merged**